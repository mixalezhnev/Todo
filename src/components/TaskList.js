import React from 'react';

import TextBox from './TextBox';
import './TaskList.res/style.css';

TaskList.propTypes = {
  children: React.PropTypes.arrayOf(React.PropTypes.element.isRequired),
  onClick: React.PropTypes.func.isRequired
}

export default function TaskList({ onClick, children }) {
  return (
    <div className="my-tasklist-component">
      <TextBox
        placeholder="Enter task title"
        onClick={onClick}/>
      <div className="tasklist">
        {children}
      </div>
    </div>
  );
}


