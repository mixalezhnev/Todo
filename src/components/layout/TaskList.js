import React from 'react';

import TextBox from '../forms/TextBox';
import './TaskList.res/style.css';

TaskList.propTypes = {
  children: React.PropTypes.arrayOf(React.PropTypes.element.isRequired),
  handleAddNewTaskClick: React.PropTypes.func.isRequired
}

export default function TaskList({ handleAddNewTaskClick, children }) {
  return (
    <div className="my-tasklist-component">
      <TextBox
        placeholder="Enter task title"
        onClick={handleAddNewTaskClick}/>
      <div className="tasklist">
        {children}
      </div>
    </div>
  );
}


