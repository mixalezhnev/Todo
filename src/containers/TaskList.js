import React, { Component, PropTypes } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import { addTask, checkDone, getTasks } from '../actions/TaskListActions';
import { getProgress } from '../actions/ProgressActions';

import TaskList from '../components/TaskList';
import TaskContainer from './Task';

class TaskListContainer extends Component {
  static propTypes = {
    showDone: PropTypes.bool.isRequired,
    tasks: PropTypes.arrayOf(PropTypes.shape({
      id: PropTypes.node.isRequired,
      title: PropTypes.string.isRequired,
      isDone: PropTypes.bool.isRequired,
      desctiptions: React.PropTypes.string.isRequired
    })),
    addTask: PropTypes.func.isRequired,
    getTasks: PropTypes.func.isRequired,
    checkDone: PropTypes.func.isRequired,
    getProgress: React.PropTypes.func.isRequired,
    URLParams: PropTypes.shape({
      id: PropTypes.string,
      category: PropTypes.string,
      filter: PropTypes.string
    })
  }

  handleAddNewTaskClick = (title) => {
    const {
      addTask,
      getProgress
    } = this.props;
    const id = this.props.URLParams.id;

    addTask(id, title);
    getProgress();
  }

  componentWillReceiveProps = (nextProps) => {
    const nextId = +nextProps.URLParams.id;
    const prevId = +this.props.URLParams.id

    if ( (typeof nextId !== undefined) && (nextId !== prevId) ){
      this.props.getTasks(nextId);
    }
  }

  _createTasks = () => {
    const {
      showDone,
      tasks
    } = this.props;
    const {
      id: categoryId,
      filter:keyWord=""
    } = this.props.URLParams;

    const newTaskList = !showDone
                        ? tasks.filter(task => !task.isDone)
                        : tasks;

    return newTaskList
            .filter((task) => task.title.toLowerCase().includes(keyWord.toLowerCase().trim()))
            .map((task, index) => {
              return (
                <TaskContainer
                  key={task.id}
                  categoryId={categoryId}
                  task={task}/>
              );
            });
  }

  render() {
    return (
      <TaskList onClick={this.handleAddNewTaskClick}>
        {this._createTasks()}
      </TaskList>
    );
  }
}

const mapStateToProps = ({ tasks }) => {
  return {
    tasks
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    addTask: bindActionCreators(addTask, dispatch),
    checkDone: bindActionCreators(checkDone, dispatch),
    getTasks: bindActionCreators(getTasks, dispatch),
    getProgress: bindActionCreators(getProgress, dispatch)
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(TaskListContainer);