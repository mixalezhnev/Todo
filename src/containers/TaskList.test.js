import React from 'react';
import { shallow, mount } from 'enzyme';

import { TaskListContainer } from './TaskList';
import TaskList from '../components/TaskList';

const setup = () => {
  const props = {
    showDone: true,
    tasks: [
      {
        id: 0,
        title: 'New Title',
        isDone: false,
        description: '1 2 34 5 6 7'
      },
      {
        id: 1,
        title: 'New second Title',
        isDone: true,
        description: '1 2 34 5 6 7'
      },
      {
        id: 3,
        title: 'New Title yeaaa',
        isDone: false,
        description: '1 2 34 5 6 7'
      },
    ],
    addTask: jest.fn(() => {
      return new Promise((resolve) => {
          resolve('I\'m born!');
        });
      }
    ),
    getProgress: jest.fn(),
    getTasks: jest.fn(),
    URLParams: {
      id: '0',
      category: 'Some category',
      filter: 'Task filter'
    }
  }
  const wrapper = shallow(<TaskListContainer {...props}/>);

  return {
    wrapper,
    props
  }
};

describe('containers', () => {
  describe('TaskListContainer', () => {
    it('should render self and subcomponents', () => {
      const { wrapper } = setup();
      
      expect(wrapper).toMatchSnapshot();
    });
    

    it('should call addTask', () => {
      const { props, wrapper } = setup();
      const newTitle = 'New title';

      wrapper.find(TaskList).prop('onClick')(newTitle);
      expect(props.addTask)
        .toBeCalledWith(props.URLParams.id, newTitle);
    });

    it('should call getTasks after receiving new props', () => {
      const { wrapper, props } = setup();
      const newProps = {
        URLParams: {
          id: 1
        }
      }
      
      wrapper.instance().componentWillReceiveProps(newProps);

      expect(props.getTasks)
        .toBeCalledWith(newProps.URLParams.id);
    });
  });
});