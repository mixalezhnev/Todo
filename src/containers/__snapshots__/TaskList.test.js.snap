exports[`containers TaskListContainer should render self and subcomponents 1`] = `
<TaskList
  onClick={[Function]} />
`;
